File: 2024-2/ejercicios/poo-sobrecarga-de-operadores/cafeteria/Cafeteria.cpp
/* 
 * File:   Cafeteria.cpp
 * Author: erichuiza
 * 
 * Created on October 6, 2024, 7:08 PM
 */

#include "Cafeteria.h"
#include "funciones.h"

Cafeteria::Cafeteria() {
    inventario = {};
    menu = {};
}

Cafeteria::Cafeteria(const Cafeteria& orig) {
}

void Cafeteria::cargarInventario(const char* nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    inventario.cargarInsumos(archivo);
}   

void Cafeteria::cargarMenu(const char* nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    menu.cargarBebidas(archivo);
}

void Cafeteria::actualizarMenu(const char* nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    while (cargarIngrediente(archivo));
    menu.verificar();
}

bool Cafeteria::cargarIngrediente(ifstream& archivo) {
    char* codigoBebida;
    char* codigoInsumo;
    double cantidad;
    char* unidadMedida;
    
    if (!archivo.eof()) {
        codigoBebida = leeCadena(archivo);
        codigoInsumo = leeCadena(archivo);
        archivo >> cantidad;
        archivo.ignore();
        unidadMedida = leeCadena(archivo, '\n');
        
        Insumo* insumo = inventario.buscarInsumo(codigoInsumo);
        Ingrediente ingrediente;
        ingrediente.setInsumo(insumo);
        ingrediente.setCantidad(cantidad);
        ingrediente.setUnidadMedida(unidadMedida);
        
        Bebida* bebida = menu.buscarBebida(codigoBebida);
        *bebida << ingrediente;
        
        return true;
    }
    
    return false;
}

void Cafeteria::reporteMenu(const char* nombreArchivo) {
    ofstream archivo(nombreArchivo, ios::out);
    archivo << menu;
}

Cafeteria::~Cafeteria() {
}

File: 2024-2/sesiones-de-clase/sesion-de-clase-23/src/Cafeteria.java
/**
 *
 * @author erichuiza
 */
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class Cafeteria {
    private Inventario inventario;
    private Menu menu;
    private Scanner archivo;
    
    public Cafeteria() {
        inventario = new Inventario();
        menu = new Menu();
        
        archivo = new Scanner(System.in);
    }
    
    public Cafeteria(String nombreArchivo) throws FileNotFoundException {
        inventario = new Inventario();
        menu = new Menu();
        
        archivo = new Scanner(new File(nombreArchivo));
    }
    
    public void cargarInventario() {
        inventario.cargarInsumos(archivo);
    }
    
    public void cargarMenu() {
        menu.cargarBebidas(archivo);
    }
    
    public void reporteMenu(final String nombreArchivo) throws IOException {
        File archivo = new File(nombreArchivo);
        FileWriter writer = new FileWriter(archivo);
        BufferedWriter buffer = new BufferedWriter(writer);

        menu.imprimir(buffer);
        writer.close();
        buffer.close();
    }
    
    public void reporteMenu() {
        menu.imprimir();
    }
}